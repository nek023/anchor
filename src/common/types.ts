export const ItemType = {
  Bookmark: 'bookmark',
  History: 'history',
  Tab: 'tab',
} as const

export interface BookmarkItem {
  type: typeof ItemType.Bookmark
  title?: string
  url?: string
}

export interface HistoryItem {
  type: typeof ItemType.History
  title?: string
  url?: string
}

export interface TabItem {
  type: typeof ItemType.Tab
  favIconUrl?: string
  title?: string
  url?: string
  tabIndex: number
  windowId: number
}

export type Item = BookmarkItem | HistoryItem | TabItem
