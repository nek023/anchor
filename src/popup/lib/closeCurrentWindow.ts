export const closeCurrentWindow = () =>
  chrome.windows.getCurrent((window) => {
    if (window.id != null) {
      chrome.windows.remove(window.id);
    }
  });
