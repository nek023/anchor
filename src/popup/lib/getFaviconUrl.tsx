// https://developer.chrome.com/docs/extensions/mv3/favicon/
export const getFaviconUrl = (pageUrl: string) => {
  const url = new URL(chrome.runtime.getURL("/_favicon/"));
  url.searchParams.set("pageUrl", pageUrl);
  url.searchParams.set("size", "32");
  return url.toString();
};
