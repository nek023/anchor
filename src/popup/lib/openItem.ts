import { Item, ItemType } from "../../shared/types";

export const openItem = (item: Item) => {
  if (item.type === ItemType.Tab) {
    chrome.tabs.highlight(
      {
        tabs: item.tabIndex,
        windowId: item.windowId,
      },
      (window) => {
        if (window.id != null) {
          chrome.windows.update(window.id, { focused: true });
        }
      },
    );
  } else {
    chrome.tabs.create({ url: item.url });
  }
};
