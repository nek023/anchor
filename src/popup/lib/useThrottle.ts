import { useEffect, useRef, useState } from "react";

// https://github.com/streamich/react-use/blob/325f5bd69904346788ea981ec18bfc7397c611df/src/useThrottle.ts
export const useThrottle = <T>(value: T, ms = 200) => {
  const [currentValue, setCurrentValue] = useState<T>(value);
  const timeout = useRef<ReturnType<typeof setTimeout>>();
  const nextValue = useRef<T>(value);
  const hasNextValue = useRef(false);

  useEffect(() => {
    if (!timeout.current) {
      setCurrentValue(value);
      const timeoutCallback = () => {
        if (hasNextValue.current) {
          hasNextValue.current = false;
          setCurrentValue(nextValue.current);
          timeout.current = setTimeout(timeoutCallback, ms);
        } else {
          timeout.current = undefined;
        }
      };
      timeout.current = setTimeout(timeoutCallback, ms);
    } else {
      nextValue.current = value;
      hasNextValue.current = true;
    }
  }, [value, ms]);

  useEffect(() => () => timeout.current && clearTimeout(timeout.current), []);

  return currentValue;
};
