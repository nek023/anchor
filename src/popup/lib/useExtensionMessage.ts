import { useEffect } from "react";
import { Message } from "../../lib/ipc";

export type ExtensionMessageCallback = (
  message: Message,
  sender: chrome.runtime.MessageSender,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  sendResponse: (response?: any) => void
) => void;

export const useExtensionMessage = (callback: ExtensionMessageCallback) => {
  useEffect(() => {
    chrome.runtime.onMessage.addListener(callback);
    return () => chrome.runtime.onMessage.removeListener(callback);
  }, [callback]);
};
