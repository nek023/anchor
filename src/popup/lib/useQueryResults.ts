import { useEffect } from "react";
import { Item } from "../../lib/types";
import { searchItems, sendMessage } from "../../lib/ipc";

export type QueryResultsCallback = (items: Item[]) => void;

export const useQueryResults = (
  query: string,
  callback: QueryResultsCallback
) => {
  useEffect(() => {
    (async () => {
      const items = (await sendMessage(searchItems(query))) as Item[];
      callback(items);
    })();
  }, [callback, query]);
};
