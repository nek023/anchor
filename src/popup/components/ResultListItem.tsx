import React, { useCallback, useEffect, useMemo, useRef } from "react";
import scrollIntoView from "scroll-into-view-if-needed";
import { Item, ItemType } from "../../shared/types";
import { getFaviconUrl } from "../lib/getFaviconUrl";

interface ResultListItemProps {
  item: Item;
  selected: boolean;
  onClick?: (item: Item) => void;
}

export const ResultListItem: React.FC<ResultListItemProps> = ({
  item,
  onClick,
  selected,
}) => {
  const itemContainerRef = useRef<HTMLDivElement>(null);

  const handleClick = useCallback(() => {
    if (onClick == null) return;
    onClick(item);
  }, [item, onClick]);

  useEffect(() => {
    if (selected && itemContainerRef?.current != null) {
      scrollIntoView(itemContainerRef?.current, {
        block: "nearest",
        inline: "nearest",
        scrollMode: "if-needed",
      });
    }
  }, [selected]);

  const faviconUrl = useMemo(() => {
    let faviconUrl: string | undefined = undefined;
    if (item.type === ItemType.Tab) faviconUrl = item.faviconUrl;
    if (faviconUrl == null && item.url != null)
      faviconUrl = getFaviconUrl(item.url);
    return faviconUrl;
  }, [item]);

  return useMemo(
    () => (
      <div
        className={`w-full min-h-[3rem] px-1 pb-px flex flex-row items-center gap-2 ${
          selected ? "bg-blue-700" : ""
        }`}
        onClick={handleClick}
        ref={itemContainerRef}
      >
        <div className="min-w-[2rem] flex justify-center">
          <img className="w-6 h-6" src={faviconUrl} alt={item.title} />
        </div>
        <div className="grow flex flex-col gap-0.5 overflow-hidden">
          <div
            className={`text-sm truncate ${
              selected ? "text-gray-100" : "text-gray-700"
            }`}
          >
            {item.title}
          </div>
          <div
            className={`text-xs font-light truncate ${
              selected ? "text-gray-300" : "text-gray-400"
            }`}
          >
            {item.url}
          </div>
        </div>
      </div>
    ),
    [faviconUrl, handleClick, item, selected],
  );
};
