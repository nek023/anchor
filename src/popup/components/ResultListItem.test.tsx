import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { BookmarkItem, ItemType } from "../../lib/types";
import { ResultListItem } from "./ResultListItem";

describe("ResultListItem", () => {
  const item: BookmarkItem = {
    id: "bookmark-1",
    type: ItemType.Bookmark,
    title: "test",
    url: "https://example.com",
  };

  test("matches snapshot", () => {
    const { asFragment } = render(
      <ResultListItem item={item} selected={false} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("calls onClick when it has been clicked", () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <ResultListItem item={item} selected={false} onClick={handleClick} />
    );

    fireEvent.click(getByText("test"));
    expect(handleClick).toBeCalledWith(item);
  });
});
