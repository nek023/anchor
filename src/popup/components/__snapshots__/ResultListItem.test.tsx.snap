// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultListItem matches snapshot 1`] = `
<DocumentFragment>
  .c1 {
  width: 30px;
  min-width: 30px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  margin-left: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c2 {
  width: 24px;
  height: 24px;
}

.c4 {
  color: #333333;
  font-size: 15px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  color: #9999aa;
  margin-top: 2px;
  font-size: 13px;
  font-weight: lighter;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c0 {
  width: 100%;
  height: 46px;
  padding: 3px;
  margin-bottom: 2px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<li
    class="c0"
  >
    <div
      class="c1"
    >
      <img
        class="c2"
        src="chrome-extension://EXTENSION_ID/_favicon/?pageUrl=https%3A%2F%2Fexample.com&size=32"
      />
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        test
      </div>
      <div
        class="c5"
      >
        https://example.com
      </div>
    </div>
  </li>
</DocumentFragment>
`;
